presto-hive/src/main/java/com/facebook/presto/hive/authentication/CachingKerberosHadoopAuthentication.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive.authentication;
+
+import com.facebook.presto.hive.HiveErrorCode;
+import com.facebook.presto.spi.PrestoException;
+import org.apache.hadoop.security.UserGroupInformation;
+
+import javax.annotation.concurrent.GuardedBy;
+import javax.security.auth.Subject;
+import javax.security.auth.kerberos.KerberosTicket;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+
+import static com.facebook.presto.hive.authentication.KerberosTicketUtils.getTgt;
+import static com.google.common.base.Preconditions.checkState;
+import static java.util.Objects.requireNonNull;
+
+public class CachingKerberosHadoopAuthentication
+        implements HadoopAuthentication
+{
+    private final KerberosHadoopAuthentication delegate;
+
+    private final Object lock = new Object();
+    @GuardedBy("lock")
+    private UserGroupInformation userGroupInformation = null;
+    @GuardedBy("lock")
+    private long nextRefreshTime = Long.MIN_VALUE;
+
+    public CachingKerberosHadoopAuthentication(KerberosHadoopAuthentication delegate)
+    {
+        this.delegate = requireNonNull(delegate, "hadoopAuthentication is null");
+    }
+
+    @Override
+    public UserGroupInformation getUserGroupInformation()
+    {
+        if (refreshIsNeeded()) {
+            synchronized (lock) {
+                if (refreshIsNeeded()) {
+                    refreshUgi();
+                }
+            }
+        }
+        return userGroupInformation;
+    }
+
+    private void refreshUgi()
+    {
+        this.userGroupInformation = delegate.getUserGroupInformation();
+        this.nextRefreshTime = calculateNextRefreshTime(this.userGroupInformation);
+    }
+
+    private boolean refreshIsNeeded()
+    {
+        return nextRefreshTime < System.currentTimeMillis() || userGroupInformation == null;
+    }
+
+    private long calculateNextRefreshTime(UserGroupInformation userGroupInformation)
+    {
+        Subject subject = getSubject(userGroupInformation);
+        checkState(subject != null, "subject must be present in kerberos based UGI");
+        KerberosTicket tgtTicket = getTgt(subject);
+        return KerberosTicketUtils.getRefreshTime(tgtTicket);
+    }
+
+    private Subject getSubject(UserGroupInformation userGroupInformation)
+    {
+        try {
+            Method getSubjectMethod = UserGroupInformation.class.getDeclaredMethod("getSubject");
```

Comment:
I replaced the reflection usages with the Shims, but as a temporary solution i put the Shims class directly to the `presto-hive` module. It will make possible to compile this branch on travis before the hadoop libraries are release.